### PR TITLE
bpo-45680: Clarify documentation on ``GenericAlias`` objects

### PR DESCRIPTION
**Pull request by AlexWaygood**
The documentation on ``GenericAlias`` objects implies at multiple points that
only container classes can define ``__class_getitem__``. This is misleading.
This PR proposes a rewrite of the documentation to clarify that non-container
classes can define ``__class_getitem__``, and to clarify what it means when a
non-container class is parameterized.

The PR also proposes some changes to the list of standard library classes that
define ``__class_getitem__``. This PR proposes adding several classes to the list,
to make it somewhat more complete. However, more importantly, it attempts
to clarify that the list is not intended to be an exhaustive enumeration of
every single standard-library class that defined ``__class_getitem__``.

See also: initial discussion of issues with this piece of documentation in
#29308, and previous BPO issue [42280](https://bugs.python.org/issue42280).

https://bugs.python.org/issue45680

### CODE REVIEW

File: Doc/library/stdtypes.rst
Comment (AlexWaygood) [inline Doc/library/stdtypes.rst:4825]: I changed the wording from "are created by subscripting a class" to "are generally created by subscripting a class". This is because it is certainly possible to directly instantiate a ``GenericAlias`` object through ``types.GenericAlias``, as this documentation states later on.

File: Doc/library/stdtypes.rst
Comment (AlexWaygood) [inline Doc/library/stdtypes.rst]: I wanted to link to a definition of the term "container", but it doesn't seem to be in the glossary.

File: Doc/library/stdtypes.rst
Comment (AlexWaygood) [inline Doc/library/stdtypes.rst]: @Fidget-Spinner, is this list currently in any particular order at all? 🙂 I can think of a few possible orderings that might work:

1. Fully alphabetise the list
2. Go through the modules in alphabetical order, but list classes within each module according to the order they appear in in the module's `__all__`.
3. Something else?

File: Doc/library/stdtypes.rst
Comment (erlend-aasland) [reply Doc/library/stdtypes.rst]:
```suggestion
container classes, such as :class:`list` or :class:`dict`. They are intended
```

File: Doc/library/stdtypes.rst
Comment (erlend-aasland) [inline Doc/library/stdtypes.rst]:
```suggestion
then the subscription of that class may instead call the class' implementation
```

File: Doc/library/stdtypes.rst
Comment (Fidget-Spinner) [reply Doc/library/stdtypes.rst]: Maybe a good one would be https://docs.python.org/3/reference/datamodel.html#emulating-container-types.

File: Doc/library/stdtypes.rst
Comment (Fidget-Spinner) [inline Doc/library/stdtypes.rst]: This example is great, but IMO we should drop it for brevity because it's already covered in the linked `__getitem__` docs.

File: Doc/library/stdtypes.rst
Comment (Fidget-Spinner) [inline Doc/library/stdtypes.rst]: Ahh, I see why you talked about ``type(food).__getitem__(food, 0)`` in such detail above. This is a nice parallel and definitely makes things easier to understand. I'd suggest extracting these bits resolving the confusion between ``__getitem__`` and ``__class_getitem__`` and place them in the docs for ``__class_getitem__`` instead https://docs.python.org/3/reference/datamodel.html#object.__class_getitem__ . I certainly found those docs somewhat lacking.

**Review by erlend-aasland — COMMENTED** [review]
I left some remarks. Generally looks good, _but_ I do not use the typing features of Python, so I'll wisely keep my mouth shut regarding most of these changes ;)

File: Doc/library/stdtypes.rst
Comment (Fidget-Spinner) [inline Doc/library/stdtypes.rst]: This feels slightly naggy. I wonder if there's a way to not repeat the ``list[int]`` example three times for separate reasons.

**Review by Fidget-Spinner — COMMENTED** [review]
I won't be able to respond so quickly due to IRL stuff. Sorry if I take a week or more!

File: Doc/library/stdtypes.rst
Comment (Fidget-Spinner) [reply Doc/library/stdtypes.rst]: Also I don't suggest writing ``They`` more than once to avoid ambiguity.

In ``They are intended``, a newbie might have no clue what's the ``They`` here refers to. ``GenericAlias`` or container classes? It doesn't hurt to repeat ``GenericAliases`` again.

**Comment by AlexWaygood** [comment]
> I won't be able to respond so quickly due to IRL stuff. Sorry if I take a week or more!

Of course, no worries! Apologies it ended up being such a large PR!

File: Doc/library/stdtypes.rst
Comment (AlexWaygood) [reply Doc/library/stdtypes.rst]: > Maybe a good one would be https://docs.python.org/3/reference/datamodel.html#emulating-container-types.

That'll do for now!

File: Doc/library/stdtypes.rst
Comment (AlexWaygood) [inline Doc/library/stdtypes.rst]: Can anybody think of a better example here? I'm not a massive fan of the example I've used, because it requires some degree of knowledge of regular expressions, which is a corner of Python not everybody knows about. (This paragraph is also longer than I'd like it to be...)

Ideally, I'd like an example that satisfies the following criteria:

1. Is a class that is relatively easy to understand. That rules out ``os.PathLike`` (it's an ABC where the significance of ``__class_getitem__`` is the return value of ``__fspath__``) and ``functools.cached_property`` (don't want to have to explain descriptors!).
2. Is a class that is *obviously* "not a container". That rules out ``queue.Queue``, which feels "kinda like a container even though it technically isn't", and anything from ``collections.abc``. (``collections.abc.Iterable`` is not a container, for example, but I don't want to use an example from ``collections.abc``, as the title of the documentation for ``collections.abc`` is literally "[collections.abc — Abstract Base Classes for Containers](https://docs.python.org/3/library/collections.abc.html)".)

If anybody has any ideas of how to improve this bit, I'd love to hear them!

File: Doc/library/stdtypes.rst
Comment (AlexWaygood) [reply Doc/library/stdtypes.rst]: I think the rule is actually slightly more complex than "never use 's'-apostrophe-'s'". The rule is generally "avoid 's'-apostrophe-'s'" with plural possessives, but not with singular possessives. (https://www.sussex.ac.uk/informatics/punctuation/apostrophe/possessives -- unless Python does something different? 🙂)

Comment (erlend-aasland) [reply Doc/library/stdtypes.rst]: Right, I learned something new today. Thanks :)

Comment (AlexWaygood) [reply Doc/library/stdtypes.rst]: Thanks for taking the time to review!! Appreciate it :)

File: Doc/library/stdtypes.rst
Comment (AlexWaygood) [inline Doc/library/stdtypes.rst:4945]: There was discussion on the previous PR that I opened on how to introduce this list. @ambv suggested "The following standard library data structures [...]" — I've gone for "The following standard library classes [...]" as I think the definition of "class" is probably easier to grok than the definition of "data structure".

File: Doc/library/stdtypes.rst
Comment (AlexWaygood) [inline Doc/library/stdtypes.rst:4838]: This material has been moved to the docs for ``__class_getitem__`` in the data model, per @Fidget-Spinner's suggestion (see #29389)

File: Doc/library/stdtypes.rst
Comment (AlexWaygood) [reply Doc/library/stdtypes.rst]: Done!

**Comment by gvanrossum** [comment]
I hope it's okay if I leave this PR in the capable hands of @Fidget-Spinner .

**Comment by Fidget-Spinner** [comment]
@AlexWaygood can I trouble you to split this PR into two please? One for `GenericAlias` and another one for `__class_getitem__`. At this point they are distinct updates to different parts of the docs ([which we commonly have separate PRs for](https://devguide.python.org/pullrequest/#making-good-commits)). You can still use the same issue number.

That will also allow me to review and accept these in smaller chunks.

**Comment by Fidget-Spinner** [comment]
@AlexWaygood  Some clarification: you don't have to close this PR, and you **do not** have to undo the changes on this branch for `__class_getitem__`, you just need to copy over those changes to a new fresh branch.

Once I accept the PRs, the merge order will be like so:
`__class_getitem__` PR -> `GenericAlias` PR.
This means it's OK for the `GenericAlias` PR to include/exclude the changes to `__class_getitem__`. I'll leave it up to you.

**Comment by AlexWaygood** [comment]
@Fidget-Spinner absolutely, I'll get on that later today.

**Comment by AlexWaygood** [comment]
> @Fidget-Spinner, I have opened 3 new pull requests [...]

I would like to file a PR addressing the docs [here](https://docs.python.org/3.11/reference/expressions.html) as well. (Though, I don't think they need to be addressed prior to this PR.)

**Comment by AlexWaygood** [comment]
> I would like to file a PR addressing the docs [here](https://docs.python.org/3.11/reference/expressions.html) as well.

First draft of the kind of thing I'm thinking of is here: #29479

**Comment by AlexWaygood** [comment]
@Fidget-Spinner - Now that #29387, #29388 and #29389 are all merged, I believe this PR and #29479 are mergeable -- I await your review! 🙂 (Whenever you have the time, of course!)

File: Doc/library/stdtypes.rst
Comment (Fidget-Spinner) [inline Doc/library/stdtypes.rst]:
```suggestion
   the special method :meth:`~object.__class_getitem__`.
```

File: Doc/library/stdtypes.rst
Comment (Fidget-Spinner) [inline Doc/library/stdtypes.rst]: This part feels a little redundant since we already mentioned the `:meth:`~object.__class_getitem__`` rule above.
```suggestion
For a container class, the
```

File: Doc/library/stdtypes.rst
Comment (Fidget-Spinner) [inline Doc/library/stdtypes.rst:4855]: Side question, I wonder if mypy knows that ``x.group(0)`` is of type `str` if we annotate ``re.Match`` with this syntax. I didn't even know the ``x.group(0)`` form was supported (but I would guess that the ``x[0]`` is).

File: Doc/library/stdtypes.rst
Comment (Fidget-Spinner) [inline Doc/library/stdtypes.rst]:
```suggestion
indicate the return type(s) of one or more methods defined on an object. For
example, :mod:`regular expressions<re>` can be used on both the :class:`str` data
type and the :class:`bytes` data type:
```

**Review by Fidget-Spinner — COMMENTED** [review]
Thanks for taking the long way around. I hope the wait wasn't too bad.

File: Doc/library/stdtypes.rst
Comment (AlexWaygood) [reply Doc/library/stdtypes.rst:4855]: [It sure does!](https://mypy-play.net/?mypy=latest&python=3.10&gist=aeb545eea714d046b845220be7ae9350)

**Comment by AlexWaygood** [comment]
> Thanks for taking the long way around. I hope the wait wasn't too bad.

Not at all -- thanks again for taking the time to review such a large PR!

**Comment by AlexWaygood** [comment]
@Fidget-Spinner, could you possibly add the `skip-news` tag on this PR and #29479, so that the CI goes green? 🙂

File: Doc/library/stdtypes.rst
Comment (Fidget-Spinner) [inline Doc/library/stdtypes.rst]: I was a little confused, the `b` eluded me, so maybe this will help
```suggestion
* If ``y = re.search(b'bar', b'bar')``, (note the `b` for :class:`bytes`), ``y`` will also be an instance
```

File: Doc/library/stdtypes.rst
Comment (Fidget-Spinner) [inline Doc/library/stdtypes.rst]: I'm going to be pedantic here: is ``re.Match[bytes]`` really a subtype of ``re.Match[Any]``? I don't remember but I'm inclined to say no.

File: Doc/library/stdtypes.rst
Comment (Fidget-Spinner) [inline Doc/library/stdtypes.rst:4866]: Off topic: I just noticed it may have been better if we originally named it `types.GenericAliasType`, similar to `types.UnionType`, but it's too late now.

**Review by Fidget-Spinner — COMMENTED** [review]
LGTM. I will ping Łukasz on Monday for a review, and I'll try to get this merged soon.

File: Doc/library/stdtypes.rst
Comment (Fidget-Spinner) [reply Doc/library/stdtypes.rst:4855]: Wow, I'm living in the past. Thanks for digging that out.

File: Doc/library/stdtypes.rst
Comment (AlexWaygood) [reply Doc/library/stdtypes.rst:4866]: Yeah. Agreed. But, as you say, `¯\_(ツ)_/¯`

File: Doc/library/stdtypes.rst
Comment (AlexWaygood) [reply Doc/library/stdtypes.rst]: I _think_ this is the correct terminology. But I'm not 100% sure, and now you mention it, I think it might be confusing wording for people who are new to type theory (sounds a bit too much like "subclass"). Maybe if we used the word "flavour" instead?

Comment (AlexWaygood) [reply Doc/library/stdtypes.rst]: Or "variety"?

Comment (Fidget-Spinner) [reply Doc/library/stdtypes.rst]: I think we can reuse the terminology in the previous paragraph, variety works good too. Good point about refraining from too much typing terminology in the stdtypes docs.

Also, I thought `re.Match` is invariant, so `re.Match[bytes]` shouldn't be a subtype of `re.Match[Any]`?

Comment (AlexWaygood) [reply Doc/library/stdtypes.rst]: Confession: I've never had any formal education in type theory (I did a music degree 😀). I was going off the definition of "subtype" [here](https://en.wikipedia.org/wiki/Subtyping):

> If S is a subtype of T, the subtyping relation is often written S <: T, to mean that any term of type S can be safely used in a context where a term of type T is expected.

mypy [is happy](https://mypy-play.net/?mypy=latest&python=3.10&flags=strict&gist=6090ffeeab4f7b0167f25ae62488761b) with us using `Match[bytes]` in any situation where `Match[Any]` is expected -- and that seems correct, to me?

But regardless, we should use different language here, we agree 😀

Comment (Fidget-Spinner) [reply Doc/library/stdtypes.rst]: > Confession: I've never had any formal education in type theory (I did a music degree 😀). I was going off the definition of "subtype" here:

Well same here (not the music degree part though), I have no clue what I'm doing most of the time 😆.

Oh, also, it seems that `typing.Match` is special, it's [generic via `typing.Match[typing.AnyStr]`](https://docs.python.org/3/library/typing.html#typing.Match), so I don't think its variance matters at all and I was probably wrong.

Thanks for the change.

**Review by Fidget-Spinner — APPROVED** [review]
LGTM. Paging @ambv for a review too please.

File: Doc/library/stdtypes.rst
Comment (TeamSpen210) [reply Doc/library/stdtypes.rst]: Perhaps `difflib.SequenceMatcher`? The generic specifies the type of item in the sequences it compares. Even if you've not used the module, the name makes fairly obvious roughly what it does.

Otherwise a not as suitable alternative is `concurrent.futures.Future`. The generic specifies the result type contained within, describing it as holding that would possibly suffice. Just not go into detail about what it's used for.

Comment (AlexWaygood) [reply Doc/library/stdtypes.rst]: TIL there's a `difflib` module! I think I'll stick with the `re` example for now — I'm happier with it now that the example's been split into bullet points, it feels more manageable — but thanks 🙂

**Comment by AlexWaygood** [comment]
I hope it's okay if I use my newfound power to add the appropriate backport labels to this PR 😉

**Comment by AlexWaygood** [comment]
Friendly ping, @ambv 🙂

File: Doc/library/stdtypes.rst
Comment (kumaraditya303) [inline Doc/library/stdtypes.rst:4827]: Container classes can be both sequence and mapping.

Comment (AlexWaygood) [reply Doc/library/stdtypes.rst:4827]: Indeed. But the link reference for container classes in general (discussing both sequences and mappings) is `sequence-types`.

Comment (AlexWaygood) [reply Doc/library/stdtypes.rst:4827]: This is where the `sequence-types` reference links to https://docs.python.org/3/reference/datamodel.html#emulating-container-types

**Comment by AlexWaygood** [comment]
Ping, @ambv, @Fidget-Spinner 🙂 any chance of getting this merged soon?

File: Doc/library/stdtypes.rst
Comment (Fidget-Spinner) [inline Doc/library/stdtypes.rst]:
```suggestion
  ``re.Match[str]``.
```

File: Doc/library/stdtypes.rst
Comment (Fidget-Spinner) [inline Doc/library/stdtypes.rst]:
```suggestion
The :meth:`~object.__getitem__` method of generic containers will raise an
exception to disallow mistakes like ``dict[str][str]``::
```

File: Doc/library/stdtypes.rst
Comment (Fidget-Spinner) [inline Doc/library/stdtypes.rst]:
```suggestion
* If ``y = re.search(b'bar', b'bar')``, (note the ``b`` for :class:`bytes`),
  ``y`` will also be an instance of ``re.Match``, but the return
  values of ``y.group(0)`` and ``y[0]`` will both be of type
  :class:`bytes`. In type annotations, we would represent this
  variety of :ref:`re.Match <match-objects>` objects with ``re.Match[bytes]``.

```

File: Doc/library/stdtypes.rst
Comment (Fidget-Spinner) [inline Doc/library/stdtypes.rst]:
```suggestion
      Introducing Python's framework for type annotations.
```

**Review by Fidget-Spinner — COMMENTED** [review]
Minor formatting suggestions, I will apply them myself then merge.

**Comment by AlexWaygood** [comment]
Hooray! Thanks, @Fidget-Spinner 😀

File: Doc/library/stdtypes.rst
Comment (Fidget-Spinner) [inline Doc/library/stdtypes.rst:5054]: Dang, I just realized the quotation marks here are inconsistent with the rest of the docs (they should be removed).

**Review by Fidget-Spinner — COMMENTED** [review]
@AlexWaygood no problem. Thanks for your patience and seeing this through.

**Pull request merged**
